File: app/src/main/java/com/blog/demo/Layer.java
package com.blog.demo;

import android.os.Parcel;
import android.os.Parcelable;

import java.util.ArrayList;
import java.util.List;

public class Layer implements Parcelable {
    private static final String SEPERATOR = "/";
    private String name;
    private String className = "";
    private List<Layer> children = new ArrayList<>();

    public Layer(String name) {
        this.name = name;
    }

    public Layer(Parcel parcel) {
        name = parcel.readString();
        className = parcel.readString();
        parcel.readTypedList(children, CREATOR);
    }

    public void addItem(String path, String className) {
        int index = path.indexOf(SEPERATOR);
        if (index == -1) {
            Layer child = addItem(path);
            child.className = className;
        } else {
            Layer child = addItem(path.substring(0, index));
            child.addItem(path.substring(index + 1), className);
        }
    }

    private Layer addItem(String path) {
        for (Layer child : children) {
            if (child.name.equals(path)) {
                return child;
            }
        }
        Layer child = new Layer(path);
        children.add(child);
        return child;
    }

    public Layer getValue(int position) {
        return children.get(position);
    }

    public int getCount() {
        return children.size();
    }

    public boolean isEmpty() {
        return children.size() == 0;
    }

    public String getName() {
        return name;
    }

    public String getClassName() {
        return className;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(name);
        dest.writeString(className);
        dest.writeTypedList(children);
    }

    public static final Creator<Layer> CREATOR = new Creator<Layer>() {
        public Layer createFromParcel(Parcel source) {
            return new Layer(source);
        }

        public Layer[] newArray(int size) {
            return new Layer[size];
        }
    };

}

File: app/src/main/java/com/blog/demo/Layer.kt
package com.blog.demo

import android.os.Parcel
import android.os.Parcelable

class Layer(var name: String?) : Parcelable {
    private val SEPERATOR = "/"

    private var className: String? = ""
    private val children: MutableList<Layer> = mutableListOf()

    constructor(parcel: Parcel) : this(parcel.readString()) {
        className = parcel.readString()
        parcel.readTypedList(children, CREATOR)
    }

    fun addItem(path: String, className: String?) {
        val index: Int = path.indexOf(SEPERATOR)
        if (index == -1) {
            val child = addItem(path)
            child.className = className
        } else {
            val child = addItem(path.substring(0, index))
            child.addItem(path.substring(index + 1), className)
        }
    }

    private fun addItem(path: String): Layer {
        for (child in children) {
            if (child.name == path) {
                return child
            }
        }
        val child = Layer(path)
        children.add(child)
        return child
    }

    fun getValue(position: Int): Layer? {
        return children[position]
    }

    fun getCount(): Int {
        return children.size
    }

    fun isEmpty(): Boolean {
        return children.isEmpty()
    }

    fun getClassName(): String? {
        return className
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeString(name)
        parcel.writeString(className)
        parcel.writeTypedList(children)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<Layer> {
        override fun createFromParcel(parcel: Parcel): Layer {
            return Layer(parcel)
        }

        override fun newArray(size: Int): Array<Layer?> {
            return arrayOfNulls(size)
        }
    }

}